dvc/scm/tree.py

```diff
@@ -58,20 +61,50 @@ def isfile(self, path):
         """Test whether a path is a regular file"""
         return os.path.isfile(path)
 
-    def walk(self, top, topdown=True, dvcignore=None):
+    def walk(self, top, topdown=True, dvcignore=None, **kwargs):
         """Directory tree generator.
 
         See `os.walk` for the docs. Differences:
         - no support for symlinks
         - it could raise exceptions, there is no onerror argument
         """
 
-        assert dvcignore
-
         def onerror(e):
             raise e
 
         for root, dirs, files in dvc_walk(
             os.path.abspath(top), dvcignore, topdown=topdown, onerror=onerror
         ):
             yield os.path.normpath(root), dirs, files
+
+
+class CleanTree(BaseTree):
+    def __init__(self, tree):
+        super(CleanTree, self).__init__()
+        self.tree = tree
+
+    @cached_property
+    def dvcignore(self):
+        return DvcIgnoreFilter(self.tree)
+
+    @property
+    def tree_root(self):
+        return self.tree.tree_root
+
+    def open(self, path, mode="r", encoding="utf-8"):
+        return self.tree.open(path, mode, encoding)
+
+    def exists(self, path):
+        return self.tree.exists(path)
+
+    def isdir(self, path):
+        return self.tree.isdir(path)
+
+    def isfile(self, path):
+        return self.tree.isfile(path)
+
+    def walk(self, top, topdown=True, **kwargs):
+        for r, d, fs in self.tree.walk(top, topdown, dvcignore=self.dvcignore):
+            d[:], fs[:] = self.dvcignore(r, d, fs)
+
+            yield r, d, fs
```

Comment:
Maybe better keep it in the dvcignore.py? Not a strong opinion, just wondering, since it is more of a wrapper than a thing that is actually related to scm.